wsdl/types.go

```diff
@@ -41,19 +56,41 @@ type Address struct {
 
 // Schema of WSDL document.
 type Schema struct {
-	XMLName      xml.Name        `xml:"schema"`
-	Imports      []*ImportSchema `xml:"import"`
-	SimpleTypes  []*SimpleType   `xml:"simpleType"`
-	ComplexTypes []*ComplexType  `xml:"complexType"`
-	Elements     []*Element      `xml:"element"`
+	XMLName         xml.Name          `xml:"schema"`
+	TargetNamespace string            `xml:"targetNamespace,attr"`
+	Namespaces      map[string]string `xml:"-"`
+	Imports         []*ImportSchema   `xml:"import"`
+	SimpleTypes     []*SimpleType     `xml:"simpleType"`
+	ComplexTypes    []*ComplexType    `xml:"complexType"`
+	Elements        []*Element        `xml:"element"`
+}
+
+// Unmarshaling solution from Matt Harden (http://grokbase.com/t/gg/golang-nuts/14bk21xb7a/go-nuts-extending-encoding-xml-to-capture-unknown-attributes)
+// We duplicate the type Schema here so that we can unmarshal into it
+// without recursively triggering the *Schema.UnmarshalXML method.
+// Other options are to embed tt into Type or declare Type as a synonym for tt.
+// The important thing is that tt is only used directly in *Schema.UnmarshalXML or Schema.MarshalXML.
+type schemaDup Schema
+
+func (schema *Schema) UnmarshalXML(d *xml.Decoder, start xml.StartElement) error {
+	for _, attr := range start.Attr {
+		if attr.Name.Space == "xmlns" {
+			if schema.Namespaces == nil {
+				schema.Namespaces = make(map[string]string)
+			}
+			schema.Namespaces[attr.Name.Local] = attr.Value
+		}
+	}
+	return d.DecodeElement((*schemaDup)(schema), &start)
 }
 
 // SimpleType describes a simple type, such as string.
 type SimpleType struct {
-	XMLName     xml.Name     `xml:"simpleType"`
-	Name        string       `xml:"name,attr"`
-	Union       *Union       `xml:"union"`
-	Restriction *Restriction `xml:"restriction"`
+	XMLName         xml.Name     `xml:"simpleType"`
+	Name            string       `xml:"name,attr"`
+	Union           *Union       `xml:"union"`
+	Restriction     *Restriction `xml:"restriction"`
+	TargetNamespace string
```

Comment:
No tag?

Reply:
This field is not for Unmarshalling. It's used for storing Namespace where XML Type is from.